core/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala

```diff
@@ -0,0 +1,207 @@
+import org.apache.spark.deploy.rest.mesos.MesosRestServer
+import org.apache.spark.scheduler.cluster.mesos._
+import org.apache.spark.util.{IntParam, SignalLogger, Utils}
+import org.apache.spark.{Logging, SecurityManager, SparkConf}
+
+/*
+ * A dispatcher that is responsible for managing and launching drivers, and is intended to
+ * be used for Mesos cluster mode. The dispatcher is launched by the user in the cluster,
+ * which it launches a [[MesosRestServer]] for listening for driver requests, and launches a
+ * [[MesosClusterScheduler]] to launch these drivers in the Mesos cluster.
+ *
+ * A typical new driver lifecycle is the following:
+ *
+ * - Driver submitted via spark-submit talking to the [[MesosRestServer]]
+ * - [[MesosRestServer]] queues the driver request to [[MesosClusterScheduler]]
+ * - [[MesosClusterScheduler]] gets resource offers and launches the drivers that are in queue
+ *
+ * This dispatcher supports both Mesos fine-grain or coarse-grain mode as the mode is configurable
+ * per driver launched.
+ * This class is needed since Mesos doesn't manage frameworks, so the dispatcher acts as
+ * a daemon to launch drivers as Mesos frameworks upon request.
+ */
+private[mesos] class MesosClusterDispatcher(
+    args: ClusterDispatcherArguments,
+    conf: SparkConf)
+  extends Logging {
+
+  private def publicAddress(conf: SparkConf, host: String): String = {
+    val envVar = conf.getenv("SPARK_PUBLIC_DNS")
+    if (envVar != null) envVar else host
+  }
+
+  private val recoveryMode = conf.get("spark.deploy.recoveryMode", "NONE").toUpperCase()
+  logInfo("Recovery mode in Mesos dispatcher set to: " + recoveryMode)
+
+  private val engineFactory = recoveryMode match {
+    case "NONE" => new BlackHoleMesosClusterPersistenceEngineFactory
+    case "ZOOKEEPER" => new ZookeeperMesosClusterPersistenceEngineFactory(conf)
+  }
+
+  private val scheduler = new MesosClusterSchedulerDriver(engineFactory, conf)
+
+  private val server = new MesosRestServer(args.host, args.port, conf, scheduler)
+  private val webUi = new MesosClusterUI(
+    new SecurityManager(conf),
+    args.webUiPort,
+    conf,
+    publicAddress(conf, args.host),
+    scheduler)
+
+  private val shutdownLatch = new CountDownLatch(1)
+  private val sparkHome = new File(Option(conf.getenv("SPARK_HOME")).getOrElse("."))
+
+  def start(): Unit = {
+    webUi.bind()
+    scheduler.frameworkUrl = webUi.activeWebUiUrl
+    scheduler.start()
+    server.start()
+  }
+
+  def awaitShutdown(): Unit = {
+    shutdownLatch.await()
+  }
+
+  def stop(): Unit = {
+    webUi.stop()
+    server.stop()
+    scheduler.stop()
+    shutdownLatch.countDown()
+  }
+}
+
+private[mesos] object MesosClusterDispatcher extends spark.Logging {
+  def main(args: Array[String]) {
+    SignalLogger.register(log)
+
+    val conf = new SparkConf
+    val dispatcherArgs = new ClusterDispatcherArguments(args, conf)
+
+    conf.setMaster(dispatcherArgs.masterUrl)
+    conf.setAppName("Spark Cluster")
```

Comment:
This is the name that is displayed on Mesos UI for the framework, I will make it configurable as well.